File: leetcode/200_number_of_islands.py
# -*- coding: utf-8 -*-

from __future__ import absolute_import, division, print_function

from collections import deque


class Solution(object):
    def numIslands(self, grid):
        if not grid:
            return 0

        result = 0

        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                if grid[i][j] == '1':
                    result += 1
                    self.deleteIsland(grid, i, j)

        return result

    def deleteIsland(self, grid, i, j):
        queue = deque([(i, j)])
        while queue:
            x, y = queue.pop()
            grid[x][y] = '0'
            if 0 < x and grid[x - 1][y] == '1':
                queue.append((x - 1, y))
            if 0 < y and grid[x][y - 1] == '1':
                queue.append((x, y - 1))
            if x < len(grid) - 1 and grid[x + 1][y] == '1':
                queue.append((x + 1, y))
            if y < len(grid[0]) - 1 and grid[x][y + 1] == '1':
                queue.append((x, y + 1))


if __name__ == '__main__':
    solution = Solution()

    assert 1 == solution.numIslands([
        ['1', '1', '1', '1', '0'],
        ['1', '1', '0', '1', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '0', '0', '0'],
    ])

    assert 3 == solution.numIslands([
        ['1', '1', '0', '0', '0'],
        ['1', '1', '0', '0', '0'],
        ['0', '0', '1', '0', '0'],
        ['0', '0', '0', '1', '1'],
    ])

File: leetcode/200_number_of_islands.kt
package leetcode

import java.util.Deque
import java.util.LinkedList

class Solution {
    fun numIslands(grid: Array<CharArray>): Int {
        var result = 0

        for (i in 0..grid.size - 1) {
            for (j in 0..grid[0].size - 1) {
                if (grid[i][j] == '1') {
                    deleteIsland(grid, i, j)
                    result++
                }
            }
        }

        return result
    }

    fun deleteIsland(grid: Array<CharArray>, i: Int, j: Int) {
        val stack: Deque<Pair<Int, Int>> = LinkedList()
        stack.addLast(Pair(i, j))

        while (stack.size > 0) {
            val (x, y) = stack.removeLast()

            grid[x][y] = '0'
            if (0 < x && grid[x - 1][y] == '1')
                stack.addLast(Pair(x - 1, y))
            if (0 < y && grid[x][y - 1] == '1')
                stack.addLast(Pair(x, y - 1))
            if (x < grid.size - 1 && grid[x + 1][y] == '1')
                stack.addLast(Pair(x + 1, y))
            if (y < grid[0].size - 1 && grid[x][y + 1] == '1')
                stack.addLast(Pair(x, y + 1))
        }
    }
}

fun main(args: Array<String>) {
    val solution = Solution()

    assert(1 == solution.numIslands(
        arrayOf(
            charArrayOf('1', '1', '1', '1', '0'),
            charArrayOf('1', '1', '0', '1', '0'),
            charArrayOf('1', '1', '0', '0', '0'),
            charArrayOf('0', '0', '0', '0', '0')
        )
    ))
    assert(3 == solution.numIslands(
        arrayOf(
            charArrayOf('1', '1', '0', '0', '0'),
            charArrayOf('1', '1', '0', '0', '0'),
            charArrayOf('0', '0', '1', '0', '0'),
            charArrayOf('0', '0', '0', '1', '1')
        )
    ))
}
